demos/evp/aesccm.c

```diff
@@ -47,79 +48,157 @@ static const unsigned char ccm_tag[] = {
     int outlen, tmplen;
     unsigned char outbuf[1024];
+
     printf("AES CCM Encrypt:\n");
     printf("Plaintext:\n");
     BIO_dump_fp(stdout, ccm_pt, sizeof(ccm_pt));
-    ctx = EVP_CIPHER_CTX_new();
-    /* Set cipher type and mode */
-    EVP_EncryptInit_ex(ctx, EVP_aes_192_ccm(), NULL, NULL, NULL);
+
+    /* Create a context for the encrypt operation */
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL)
+        goto err;
+
+    /* Fetch the cipher implementation */
+    if ((cipher = EVP_CIPHER_fetch(NULL, "AES-192-CCM", NULL)) == NULL)
+        goto err;
+
+    /* Initialise encrypt operation with the cipher & mode */
+    if (!EVP_EncryptInit_ex(ctx, cipher, NULL, NULL, NULL))
+        goto err;
+
     /* Set nonce length if default 96 bits is not appropriate */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
-                        NULL);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
+                             NULL))
+        goto err;
+
     /* Set tag length */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, sizeof(ccm_tag), NULL);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, sizeof(ccm_tag),
+                             NULL))
+        goto err;
+
     /* Initialise key and IV */
-    EVP_EncryptInit_ex(ctx, NULL, NULL, ccm_key, ccm_nonce);
+    if (!EVP_EncryptInit_ex(ctx, NULL, NULL, ccm_key, ccm_nonce))
+        goto err;
+
     /* Set plaintext length: only needed if AAD is used */
-    EVP_EncryptUpdate(ctx, NULL, &outlen, NULL, sizeof(ccm_pt));
+    if (!EVP_EncryptUpdate(ctx, NULL, &outlen, NULL, sizeof(ccm_pt)))
+        goto err;
+
     /* Zero or one call to specify any AAD */
-    EVP_EncryptUpdate(ctx, NULL, &outlen, ccm_adata, sizeof(ccm_adata));
+    if (!EVP_EncryptUpdate(ctx, NULL, &outlen, ccm_adata, sizeof(ccm_adata)))
+        goto err;
+
     /* Encrypt plaintext: can only be called once */
-    EVP_EncryptUpdate(ctx, outbuf, &outlen, ccm_pt, sizeof(ccm_pt));
+    if (!EVP_EncryptUpdate(ctx, outbuf, &outlen, ccm_pt, sizeof(ccm_pt)))
+        goto err;
+
     /* Output encrypted block */
     printf("Ciphertext:\n");
     BIO_dump_fp(stdout, outbuf, outlen);
+
     /* Finalise: note get no output for CCM */
-    EVP_EncryptFinal_ex(ctx, outbuf, &outlen);
+    if (!EVP_EncryptFinal_ex(ctx, outbuf, &outlen))
+        goto err;
+
     /* Get tag */
-    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outbuf);
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outbuf))
+        goto err;
+
     /* Output tag */
     printf("Tag:\n");
     BIO_dump_fp(stdout, outbuf, 16);
+
+    ret = 1;
+err:
+    if (!ret)
+        ERR_print_errors_fp(stderr);
+
+    EVP_CIPHER_free(cipher);
     EVP_CIPHER_CTX_free(ctx);
+
+    return ret;
 }
 
-void aes_ccm_decrypt(void)
+int aes_ccm_decrypt(void)
 {
+    int ret = 0;
     EVP_CIPHER_CTX *ctx;
+    EVP_CIPHER *cipher;
     int outlen, tmplen, rv;
     unsigned char outbuf[1024];
+
```

Comment:
should we use libctx and propq in here.. 
They could be variables set to NULL? with a comment saying they can be set if required?

Reply:
Yes!

Reply:
I've added a comment and globals for libctx/propq which are passed to EVP_CIPHER_fetch().